server/market/bookrouter.go

```diff
@@ -0,0 +1,389 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package market
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	"github.com/decred/dcrdex/server/comms"
+	"github.com/decred/dcrdex/server/comms/msgjson"
+	"github.com/decred/dcrdex/server/market/types"
+	"github.com/decred/dcrdex/server/order"
+)
+
+// An update action classifies updates into how they affect the book or epoch
+// queue.
+type updateAction uint8
+
+const (
+	// epochAction means an order is being added to the epoch queue and will
+	// result in a msgjson.EpochOrderNote being sent to subscribers.
+	epochAction updateAction = iota
+	// bookAction means an order is being added to the order book, and will result
+	// in a msgjson.BookOrderNote being sent to subscribers.
+	bookAction
+	// unbookAction means an order is being removed from the order book and will
+	// result in a msgjson.UnbookOrderNote being sent to subscribers.
+	unbookAction
+)
+
+// updateSignal combines an updateAction with the order on which the action
+// applies.
+type updateSignal struct {
```

Comment:
The `Market` will have this as a field, to which it sends order book updates.  But since there will be a similar channel (or something else) for incoming orders from the order router planned in https://github.com/decred/dcrdex/issues/52, perhaps we can name `updateSignal` something less ambiguous.

Reply:
Probably something along these lines

```go
type Market struct {
	running bool

	// Communications.
	orderRouter <-chan *orderUpdateSignal // incoming orders
	orderFeed   chan<- *bookUpdateSignal  // outgoing notifications
```

Reply:
That sounds good. A couple of questions.

1. Who makes the channels? Is it going to be `Market` that satisfies `BookSource`?

2. Does the `BookSource` just have one channel, and repeated calls to `OrderFeed` return the same `chan`, or does the `BookSource` support multiple feeders?

Reply:
If for no other reason than `Market` being the _sender_ on the channel, only `Market` should close the channel.  This probably means it should `make` it too, but it could be passed down from the `DEX` when it creates the `Market`.

A guess at what DEX would do:
- create `Market`s, which creates the order book subscription update channel
- get books or `BookSources` from markets
- create the `NewOrderBookRouter` with the `BookRouterConfig` built from all of the `BookSources`

Reply:
Oh, at this point I can't think of a reason by BookSource would need multiple feeders.